### PR TITLE
Code updated from data-source-juggler

### PR DESCRIPTION
**Pull request by privateOmega**
### Description
Code refactoring and updates from data-source juggler

#### Related issues
- connect to [datasource-juggler#1529 ](https://github.com/strongloop/loopback-datasource-juggler/issues/1529)
- connect to [#13 ](https://github.com/strongloop/loopback-filters/issues/13)
- connect to [#10 ](https://github.com/strongloop/loopback-filters/issues/10#issuecomment-351646274)

### Checklist
- [ ] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style
  guide](http://loopback.io/doc/en/contrib/style-guide.html)

### CODE REVIEW

**Comment by privateOmega** [comment]
@kjdelisle @b-admike @virkt25 @dhmlau @jannyHou Please review the PR.

**Comment by privateOmega** [comment]
@bajtos Kindly review this PR to close it

**Review by bajtos — CHANGES_REQUESTED** [review]
@privateOmega thank you for the pull request and sorry for the long silence at our side!

IIUC, your pull request is existing porting code from loopback-datasource-juggler. To make future maintenance easier and prevent regressions, we need you to port the tests as well.

Also, while you are porting the code, please update it to use `const` instead of `var`, see http://loopback.io/doc/en/contrib/style-guide.html#variable-declarations

File: index.js [hunk 1]
```diff
@@ -6,7 +6,12 @@
 var debug = require('debug')('loopback:filter');
 var geo = require('./lib/geo');
 
-module.exports = function filterNodes(nodes, filter) {
+module.exports = {
+  filterNodes: filterNodes,
+  applyFilter: applyFilter,
+};
```
Comment (bajtos) [inline index.js]: This is a breaking change I'd rather avoid.

I am proposing to keep `filterNodes` as the main exported function, and add `applyFilter` as a property of this function.

```js
module.exports = filterNodes;
filterNodes.applyFilter = applyFilter;
```

Thoughts?

Regardless of what approach we decide to take, README needs to be updated to show usage of `applyFilter` too.

File: index.js [hunk 2]
```diff
@@ -157,15 +192,7 @@ function test(example, value) {
...
-      var regexp;
-      if (example.regexp instanceof RegExp) {
-        regexp = example.regexp;
-      } else if (typeof example.regexp === 'string') {
-        regexp = toRegExp(example.regexp);
-      } else {
-        throw new TypeError('Invalid regular expression passed to regexp query.');
-      }
-      return regexp.test(value);
+      return value ? value.match(RegExp(example.regexp)) : false;
```
Comment (bajtos) [inline index.js]: Isn't this a breaking change too? I mean before your change, when the `regexp` value was not a string or RegExp instance, the code thrown a reasonably helpful error. With your change in place, we will try to convert anything to a regexp, e.g. `RegExp(0)` or `RegExp(new Date())`.

I think we should preserve the old code.

What are your arguments for making this change?

File: lib/geo.js [hunk 3]
```diff
@@ -11,75 +11,108 @@ var assert = require('assert');
...
+      // filter result if distance is either < minDistance or > maxDistance
+      if ((min && d < min) || (max && d > max)) return;
+      distances[result.id] = d;
+      results.push(result);
+    });
 
-  return result.sort(function(objA, objB) {
-    var a = objB[key];
-    var b = objB[key];
+    results.sort(function(resA, resB) {
```
Comment (bajtos) [inline lib/geo.js]: Could you please preserve argument names `objA` and `objB` to make the diff smaller and easier to review?

File: test/filter.test.js [hunk 4]
```diff
@@ -5,7 +5,7 @@
 'use strict';
 var assert = require('assert');
 var should = require('should');
-var filter = require('../');
+var filter = require('../index.js').filterNodes;
```
Comment (bajtos) [inline test/filter.test.js]: I think `require('../').filterNode` is better, no need to rely on a specific `main` file.

File: index.js (hunk 1 above)
Comment (privateOmega) [reply index.js]: Sure. Will definitely update the README as well.

**Comment by bajtos** [comment]
Hello, what's the status of this pull request? I see there was a new commit added after my comments, is the patch ready for another round of review?

**Comment by bajtos** [comment]
Closing as abandoned.

**Pull request closed without merging**